proxy/proxy.go

```diff
@@ -476,6 +479,23 @@ func (p *Proxy) Resolve(d *DNSContext) error {
 	return err
 }
 
+// Resolve and update the DNS cache.
+func (p *Proxy) resolveAndUpdateCache(d *DNSContext) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Error("resolveAndUpdateCache: recovered from panic: %v", r)
+		}
+	}()
+	d2 := &DNSContext{
+		Req:                 d.Req.Copy(),
+		disableCacheReading: true,
+	}
+	err := p.Resolve(d2)
+	if err != nil {
+		log.Debug("Resolve failed %s", err)
```

Comment:
One more thing that I just noticed.

I am not sure we should serve expired records in case we cannot resolve them anymore.

I guess it'd be better to simple remove it from the cache.